From cf726c71c93fb0525d86b325891be8e271d471d5 Mon Sep 17 00:00:00 2001
From: Flavio Paiva Junqueira <fpj@apache.org>
Date: Sat, 7 Dec 2013 10:17:54 +0000
Subject: [PATCH 61/94] ZOOKEEPER-1459. Standalone ZooKeeperServer is not
 closing   the transaction log files on shutdown
 (Rakesh R via fpj)

git-svn-id: https://svn.apache.org/repos/asf/zookeeper/branches/branch-3.4@1548826 13f79535-47bb-0310-9956-ffa450edef68

Conflicts:
	CHANGES.txt
---
 CHANGES.txt                                        |    3 +++
 .../zookeeper/server/ZooKeeperServerMain.java      |   11 +++++++---
 .../zookeeper/server/ZooKeeperServerMainTest.java  |   23 ++++++++++++++++++--
 3 files changed, 32 insertions(+), 5 deletions(-)

diff --git a/CHANGES.txt b/CHANGES.txt
index 888fcfa..8d97696 100644
--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -87,6 +87,9 @@ BUGFIXES:
   ZOOKEEPER-1557. jenkins jdk7 test failure in
   testBadSaslAuthNotifiesWatch (Eugene Koontz via phunt)
 
+  ZOOKEEPER-1459. Standalone ZooKeeperServer is not closing 
+  the transaction log files on shutdown (Rakesh R via fpj)
+
 IMPROVEMENTS:
 
 
diff --git a/src/java/main/org/apache/zookeeper/server/ZooKeeperServerMain.java b/src/java/main/org/apache/zookeeper/server/ZooKeeperServerMain.java
index 48bc2c5..9d09cdb 100644
--- a/src/java/main/org/apache/zookeeper/server/ZooKeeperServerMain.java
+++ b/src/java/main/org/apache/zookeeper/server/ZooKeeperServerMain.java
@@ -98,6 +98,7 @@ public class ZooKeeperServerMain {
      */
     public void runFromConfig(ServerConfig config) throws IOException {
         LOG.info("Starting server");
+        FileTxnSnapLog txnLog = null;
         try {
             // Note that this thread isn't going to be doing anything else,
             // so rather than spawning another thread, we will just call
@@ -105,9 +106,9 @@ public class ZooKeeperServerMain {
             // create a file logger url from the command line args
             ZooKeeperServer zkServer = new ZooKeeperServer();
 
-            FileTxnSnapLog ftxn = new FileTxnSnapLog(new
-                   File(config.dataLogDir), new File(config.dataDir));
-            zkServer.setTxnLogFactory(ftxn);
+            txnLog = new FileTxnSnapLog(new File(config.dataLogDir), new File(
+                    config.dataDir));
+            zkServer.setTxnLogFactory(txnLog);
             zkServer.setTickTime(config.tickTime);
             zkServer.setMinSessionTimeout(config.minSessionTimeout);
             zkServer.setMaxSessionTimeout(config.maxSessionTimeout);
@@ -122,6 +123,10 @@ public class ZooKeeperServerMain {
         } catch (InterruptedException e) {
             // warn, but generally this is ok
             LOG.warn("Server interrupted", e);
+        } finally {
+            if (txnLog != null) {
+                txnLog.close();
+            }
         }
     }
 
diff --git a/src/java/test/org/apache/zookeeper/server/ZooKeeperServerMainTest.java b/src/java/test/org/apache/zookeeper/server/ZooKeeperServerMainTest.java
index e3f6dc2..d276059 100644
--- a/src/java/test/org/apache/zookeeper/server/ZooKeeperServerMainTest.java
+++ b/src/java/test/org/apache/zookeeper/server/ZooKeeperServerMainTest.java
@@ -47,10 +47,11 @@ public class ZooKeeperServerMainTest extends ZKTestCase implements Watcher {
     public static class MainThread extends Thread {
         final File confFile;
         final TestZKSMain main;
+        final File tmpDir;
 
         public MainThread(int clientPort) throws IOException {
             super("Standalone server with clientPort:" + clientPort);
-            File tmpDir = ClientBase.createTmpDir();
+            tmpDir = ClientBase.createTmpDir();
             confFile = new File(tmpDir, "zoo.cfg");
 
             FileWriter fwriter = new FileWriter(confFile);
@@ -89,9 +90,25 @@ public class ZooKeeperServerMainTest extends ZKTestCase implements Watcher {
             }
         }
 
-        public void shutdown() {
+        public void shutdown() throws IOException {
             main.shutdown();
         }
+
+        void deleteDirs() throws IOException{
+            delete(tmpDir);
+        }
+
+        void delete(File f) throws IOException {
+            if (f.isDirectory()) {
+                for (File c : f.listFiles())
+                    delete(c);
+            }
+            if (!f.delete())
+                // double check for the file existence
+                if (f.exists()) {
+                    throw new IOException("Failed to delete file: " + f);
+                }
+        }
     }
 
     public static  class TestZKSMain extends ZooKeeperServerMain {
@@ -126,6 +143,8 @@ public class ZooKeeperServerMainTest extends ZKTestCase implements Watcher {
         zk.close();
 
         main.shutdown();
+        main.join();
+        main.deleteDirs();
 
         Assert.assertTrue("waiting for server down",
                 ClientBase.waitForServerDown("127.0.0.1:" + CLIENT_PORT,
-- 
1.7.9.5

